ReactiveCocoa/Swift/Flatten.swift

```diff
@@ -0,0 +1,571 @@
+}
+
+private final class ConcatState<Value, Error: ErrorType> {
+	/// The observer of a started `concat` producer.
+	let observer: Observer<Value, Error>
+
+	/// The top level disposable of a started `concat` producer.
+	let disposable: CompositeDisposable?
+
+	/// The active producer, if any, and the producers waiting to be started.
+	let queuedSignalProducers: Atomic<[SignalProducer<Value, Error>]> = Atomic([])
+
+	init(observer: Signal<Value, Error>.Observer, disposable: CompositeDisposable?) {
+		self.observer = observer
+		self.disposable = disposable
+	}
+
+	func enqueueSignalProducer(producer: SignalProducer<Value, Error>) {
+		if let d = disposable where d.disposed {
+			return
+		}
+
+		var shouldStart = true
+
+		queuedSignalProducers.modify { (var queue) in
+			// An empty queue means the concat is idle, ready & waiting to start
+			// the next producer.
+			shouldStart = queue.isEmpty
+			queue.append(producer)
+			return queue
+		}
+
+		if shouldStart {
+			startNextSignalProducer(producer)
+		}
+	}
+
+	func dequeueSignalProducer() -> SignalProducer<Value, Error>? {
+		if let d = disposable where d.disposed {
+			return nil
+		}
+
+		var nextSignalProducer: SignalProducer<Value, Error>?
+
+		queuedSignalProducers.modify { (var queue) in
+			// Active producers remain in the queue until completed. Since
+			// dequeueing happens at completion of the active producer, the
+			// first producer in the queue can be removed.
+			if !queue.isEmpty { queue.removeAtIndex(0) }
+			nextSignalProducer = queue.first
+			return queue
+		}
+
+		return nextSignalProducer
+	}
+
+	/// Subscribes to the given signal producer.
+	func startNextSignalProducer(signalProducer: SignalProducer<Value, Error>) {
+		signalProducer.startWithSignal { signal, disposable in
+			let handle = self.disposable?.addDisposable(disposable) ?? nil
+
+			signal.observe { event in
+				switch event {
+				case .Completed, .Interrupted:
+					handle?.remove()
+
+					if let nextSignalProducer = self.dequeueSignalProducer() {
+						self.startNextSignalProducer(nextSignalProducer)
+					}
+
+				default:
+					self.observer.action(event)
+				}
+			}
+		}
+	}
+}
+
+extension SignalType where Value: SignalProducerType, Error == Value.Error {
+	/// Merges a `signal` of SignalProducers down into a single signal, biased toward the producer
+	/// added earlier. Returns a Signal that will forward events from the inner producers as they arrive.
+	@warn_unused_result(message="Did you forget to call `start` on the producer?")
+	private func merge() -> Signal<Value.Value, Error> {
+		return Signal<Value.Value, Error> { relayObserver in
+			self.observeMerge(MergeState(observer: relayObserver, disposable: nil))
+		}
+	}
+
+	private func observeMerge(state: MergeState<Value.Value, Error>) -> Disposable? {
+		return self.observe { event in
+			switch event {
+			case let .Next(producer):
+				producer.startWithSignal { innerSignal, innerDisposable in
+					state.inFlight.modify { $0 + 1 }
```

Comment:
While I appreciate your suggestion to improve this, I think it would be great to leave that for a separate change so that this is purely rearranging the existing implementation. Otherwise we risk potentially not knowing what broke this operator.